### PR TITLE
neovim: suffix path with extraPackages

### PR DESCRIPTION
**Pull request by teto**
### Description

@jlesquembre  like this ?
Fixes https://github.com/nix-community/home-manager/issues/1739 

### Checklist

- [ ] Change is backwards compatible.

- [ ] Code formatted with `./format`.

- [ ] Code tested through `nix-shell --pure tests -A run.all`.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [ ] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/doc/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

File: modules/programs/neovim.nix [hunk 1]
```diff
@@ -43,7 +43,7 @@ let
 
   # A function to get the configuration string (if any) from an element of 'plugins'
   pluginConfig = p:
-    if builtins.hasAttr "plugin" p && builtins.hasAttr "config" p then ''
+    if builtins.hasAttr "plugin" p && p.config or "" != "" then ''
```
Comment (jlesquembre) [inline modules/programs/neovim.nix:46]: Aren't both versions equivalent? I find more readable the `hasAttr` version

**Comment by jlesquembre** [comment]
@teto yes, that's what I have in mind, looks good, but I'm not sure about the change in line 46. Is it necessary?

File: modules/programs/neovim.nix (hunk 1 above)
Comment (berbiche) [reply modules/programs/neovim.nix:46]: An for the shortest version: `p ? plugin && p ? config`

**Comment by berbiche** [comment]
I am ambivalent to this solution. If the `extraPackages` are appended to the `$PATH` then whatever is currently present in the `$PATH` trumps the `extraPackages` and if they are prepended then `direnv`'s magic stops working.

`$PATH` is such an annoyance.
There is no ideal solution.

I personally would keep the current behavior of prepending and in @jlesquembre case, write an alias or dummy wrapper around neovim with custom `$PATH` logic.

**Comment by jlesquembre** [comment]
@berbiche agree that append or prepend to `PATH` depends on your use case, but I still think that in this case is more intuitive if we append to  `PATH`.
While I mentioned `direnv`, it's the same if you use `nix-shell`. I think is confusing to a have a dependency in `extraPackages` that is not overridden when you use a nix shell. Keep in mind that those binaries are not called directly by vim, but by the user, via the `!` command, the terminal emulator or a plugin.
IMO, `extraPackages` is similar to `home.packages`, but for vim, and the behavior should be same. 

Another possibility could be to add a new option, `defaultPackages`, and append to `PATH` in these case. But I'm not sure if is good idea to add a new option.

**Comment by berbiche** [comment]
I agree with all your points.
My comprehension of this option was that it added packages to the neovim environment that plugins would need and as such it made more sense to me to have these packages prepended to the `$PATH`.
If that is not the common usage, then I agree with the changes.

**Comment by dotlambda** [comment]
I have only `vimAlias` and `configure` set and the assertion is triggering. I think this change is at fault.

**Comment by teto** [comment]
we could get rid of configure: is there anything that can be done with `configure` that can t be done via the other home-manager options ?

**Comment by dotlambda** [comment]
> is there anything that can be done with `configure` that can t be done via the other home-manager options ?

I have a problem loading plugins when I don't use `plug`:
```
Error detected while processing /home/rschuetz/.config/nvim/init.vim:
line   30:
E121: Undefined variable: g:vimtex#re#deoplete
E116: Invalid arguments for function deoplete#custom#var
E325: ATTENTION
```
Maybe the order in which plugins are loaded matters, but deoplete comes before vimtex in my `plugins` list.

**Comment by dotlambda** [comment]
> we could get rid of configure

Also, I think breaking something isn't a good reason for removing the feature. It should be fixed.

**Comment by teto** [comment]
> Also, I think breaking something isn't a good reason for removing the feature

configure creates complexity and is an historical artefact of the way things were done in nixpkgs, Other home-manager options are typed and superior to `configure` so it would be a good opportunity to get rid of it.

**Comment by dotlambda** [comment]
> so it would be a good opportunity to get rid of it

Sure, if everything worked with the new options, I'd be happy to switch away from `configure`.

**Comment by teto** [comment]
cool let me know if anything is missing.

**Comment by dotlambda** [comment]
> cool let me know if anything is missing.

See https://github.com/nix-community/home-manager/pull/1756#issuecomment-774633369.

**Comment by teto** [comment]
ha sry somehow I missed that, are your dotfiles/neovim config public ?

**Comment by dotlambda** [comment]
https://0x0.st/-Hji.nix

**Comment by dotlambda** [comment]
@teto Do you have any idea?

**Comment by teto** [comment]
deoplete tries to use a variable that was not declared yet. I hadn't have the chance to test yet but maybe try to reorder the list so that vimtex appear before deoplete. I dont think that should influcence the order since plugins should be loaded in same order but let's try. Otherwise, make your deoplete plugin optional and load it afterwards with `packadd! deoplete` or sthg similar.

**Comment by teto** [comment]
@dotlambda any chance with my propositons ?

**Comment by dotlambda** [comment]
Reordering has no effect.
Making deoplete optional gives
```
E919: Directory not found in 'packpath': "pack/*/opt/deoplete"
```
and the error
```
E121: Undefined variable: g:vimtex#re#deoplete
```
remains. Hence I think it's more likely to be a problem with loading vimtex.

**Comment by teto** [comment]
2nd error is a consequence of `E919: Directory not found in 'packpath': "pack/*/opt/deoplete"`.  Thats strange. Are you able to find out why ? what does. `:verbose set packpath?` say ? is your ~/.config/nvim/init.vim properly generated ?

**Comment by dotlambda** [comment]
> 2nd error is a consequence of `E919: Directory not found in 'packpath': "pack/*/opt/deoplete"`.

I doubt it. I'm pretty sure `g:vimtex#re#deoplete` is defined by vimtex. If the problem was with deoplete, it should rather be `deoplete#custom#var` that's undefined.

**Comment by dotlambda** [comment]
My final fix to make this work was to wrap the call of `deoplete#custom#var` in an `autocmd FileType tex`. I guess it's related to https://github.com/lervag/vimtex/issues/1413.

**Pull request merged**